### PR TITLE
[Hexagon][LLVM] Enable/test tensorized Hexagon DMA on 2d transformed layout

### PR DESCRIPTION
**Pull request by Lunderberg**
- In the `CodeGenLLVM::CreateIntrinsic` handler for `builtin::address_of()`, pass N-d indices to `CodeGenLLVM::CreateBufferPtr`.  The base class implementation still asserts that there is a flat memory space, while the `CodeGenHexagon::CreateBufferPtr` override allows 2-d memory.

- Enable tensorization in `test_cache_read_write.py`, using `tir.address_of` to pass the lowered value.

Co-authored-by: Adam Straw <astraw@octoml.ai>

### CODE REVIEW

File: tests/python/contrib/test_hexagon/test_cache_read_write.py [hunk 1]
```diff
@@ -38,33 +37,49 @@ def intrin_mem_copy(shape, dtype, dst_scope, src_scope):
...
 
     return te.decl_tensor_intrin(dst.op, intrin_func, binds={src: src_buffer, dst: dst_buffer})
 
 
+def layout_transform_2d(n):
+    return [n // 16, te.AXIS_SEPARATOR, n % 16]
+
+
 @requires_hexagon_toolchain
 def test_cache_read_write(hexagon_session):
```
Comment (csullivan) [inline tests/python/contrib/test_hexagon/test_cache_read_write.py:81]: Can we have a test demonstrating the approach for discontiguous memory _and_ contiguous memory? I notice you are overwriting the old test coverage with this change and it's likely useful to maintain coverage for both cases.

File: src/tir/ir/buffer.cc [hunk 2]
```diff
@@ -480,8 +479,14 @@ Buffer Buffer::MakeSlice(Array<PrimExpr> begins, Array<PrimExpr> extents) const
...
     }
   }
-  return Buffer(n->data, n->dtype, extents, strides, elem_offset[0], n->name + "_slice",
-                n->data_alignment, 0, n->buffer_type);
+  Buffer slice(n->data, n->dtype, extents, strides, elem_offset[0], n->name + "_slice",
+               n->data_alignment, 0, n->buffer_type);
+  if (elem_offset.size() != 1) {
+    // Sentinel value for ArgBinder::BindBuffer to state that any usage
+    // of element offset is invalid.
+    slice.CopyOnWrite()->elem_offset = PrimExpr();
```
Comment (csullivan) [inline src/tir/ir/buffer.cc:487]: Expanding this comment to explain why the use of element offset is invalid in the Nd case, or even better a short TODO to update once the IR/buffer is changed to support Nd offset would help. It took me a little while to understand why this was necessary.

File: tests/python/contrib/test_hexagon/test_cache_read_write.py (hunk 1 above)
Comment (adstraw) [reply tests/python/contrib/test_hexagon/test_cache_read_write.py]: Done

File: src/tir/ir/buffer.cc (hunk 2 above)
Comment (adstraw) [reply src/tir/ir/buffer.cc]: @Lunderberg Let me know if you are OK with the comment rewrite here including the `TODO` I wrote related to PR #10816.

**Review by csullivan — APPROVED** [review]

**Comment by csullivan** [comment]
Thanks @adstraw @Lunderberg! This is merged

**Pull request merged**